tools/serverpod_cli/test/database/create_definition/database_action_definition_test.dart

```diff
@@ -0,0 +1,108 @@
+import 'package:test/test.dart';
+
+void main() {
+  test(
+      'Given a class definition with a table, then generate a table with that name.',
+      () {
+    var field = FieldDefinitionBuilder().withPrimaryKey().build();
+    var entity = ClassDefinitionBuilder()
+        .withTableName('example')
+        .withField(field)
+        .build();
+
+    var databaseDefinition = createDatabaseDefinitionFromEntities([entity]);
+
+    expect(databaseDefinition.tables, hasLength(1));
+    expect(databaseDefinition.tables.first.name, 'example');
+  });
+  group(
+      'Given a class definition with a foreign relation with onDelete and onUpdate set to "SetNull"',
+      () {
+    var relation = ForeignRelationDefinitionBuilder()
+        .withParentTable('example')
+        .withReferenceFieldName('id')
+        .withOnDelete(ForeignKeyAction.setNull)
+        .withOnUpdate(ForeignKeyAction.setNull)
+        .build();
+
+    var field = FieldDefinitionBuilder()
+        .withName('parentId')
+        .withIdType(true)
+        .withRelation(relation)
+        .build();
+
+    var entity = ClassDefinitionBuilder()
+        .withTableName('example')
+        .withField(field)
+        .build();
+
+    var databaseDefinition = createDatabaseDefinitionFromEntities([entity]);
+
+    var failedPrecondition = databaseDefinition.tables.isEmpty;
+    test('then a foreign relation exists.', () {
+      var table = databaseDefinition.tables.first;
+      expect(
+        table.foreignKeys,
+        isNotEmpty,
+        reason: 'Expected a foreign relation to exists.',
+      );
+    }, skip: failedPrecondition);
+
+    group(' ', () {
+      var table = databaseDefinition.tables.first;
+
+      var failedPrecondition =
+          databaseDefinition.tables.first.foreignKeys.isEmpty;
+
+      test('then the foreign key references is a self reference.', () {
+        var foreignKey = table.foreignKeys.first;
+
+        var referenceDatabase = foreignKey.referenceTable;
+        expect(referenceDatabase, 'example');
+
+        var referenceColumn = foreignKey.referenceColumns.first;
+        expect(referenceColumn, 'id');
+      }, skip: failedPrecondition);
+
+      test(
+          'then generate a database definition with onDelete set on the foreign key.',
+          () {
+        var foreignKey = table.foreignKeys.first;
+
+        expect(foreignKey.onDelete, ForeignKeyAction.setNull);
+      }, skip: failedPrecondition);
+
+      test(
+          'then generate a database definition with onUpdate set on the foreign key.',
+          () {
+        var foreignKey = table.foreignKeys.first;
+        expect(foreignKey.onUpdate, ForeignKeyAction.setNull);
+      }, skip: failedPrecondition);
+    }, skip: failedPrecondition);
+
+    group('when generating sql code', () {
+      // TODO: fix this, not sure what this does or why it is not set in the
+      // definition creator ?! But code generator crashes without it.
```

Comment:
I found this issue and hot patched it in a test as well, seems to be used to define in what order migrations should be applied.

I think we should maybe fix this properly so that we don't access a nullable value unconditionally.

Reply:
In a separate PR. Leets look at it together.